miniapps/navier/navier_solver.cpp

```diff
@@ -0,0 +1,1048 @@
+      dynamic_cast<HypreSmoother *>(MvInvPC)->SetType(HypreSmoother::Jacobi, 1);
+   }
+   MvInv = new CGSolver(MPI_COMM_WORLD);
+   MvInv->iterative_mode = false;
+   MvInv->SetOperator(*Mv);
+   MvInv->SetPreconditioner(*MvInvPC);
+   MvInv->SetPrintLevel(pl_mvsolve);
+   MvInv->SetRelTol(1e-12);
+   MvInv->SetMaxIter(200);
+
+   if (partial_assembly)
+   {
+      Sp_form_lor = new ParBilinearForm(pfes_lor);
+      Sp_form_lor->UseExternalIntegrators();
+      CopyDBFIntegrators(Sp_form, Sp_form_lor);
+      Sp_form_lor->Assemble();
+      Sp_form_lor->FormSystemMatrix(pres_ess_tdof, Sp_lor);
+      SpInvPC = new HypreBoomerAMG(*Sp_lor.As<HypreParMatrix>());
+      SpInvPC->SetPrintLevel(pl_amg);
+      SpInvPC->Mult(resp, pn);
+      SpInvOrthoPC = new OrthoSolver();
+      SpInvOrthoPC->SetOperator(*SpInvPC);
+   }
+   else
+   {
+      SpInvPC = new HypreBoomerAMG(*Sp.As<HypreParMatrix>());
+      SpInvPC->SetPrintLevel(0);
+      SpInvOrthoPC = new OrthoSolver();
+      SpInvOrthoPC->SetOperator(*SpInvPC);
+   }
+   SpInv = new CGSolver(MPI_COMM_WORLD);
+   SpInv->iterative_mode = true;
+   SpInv->SetOperator(*Sp);
+   if (pres_dbcs.empty())
+   {
+      SpInv->SetPreconditioner(*SpInvOrthoPC);
+   }
+   else
+   {
+      SpInv->SetPreconditioner(*SpInvPC);
+   }
+   SpInv->SetPrintLevel(pl_spsolve);
+   SpInv->SetRelTol(rtol_spsolve);
+   SpInv->SetMaxIter(200);
+
+   if (partial_assembly)
+   {
+      Vector diag_pa(vfes->GetTrueVSize());
+      H_form->AssembleDiagonal(diag_pa);
+      HInvPC = new OperatorJacobiSmoother(diag_pa, vel_ess_tdof);
+   }
+   else
+   {
+      HInvPC = new HypreSmoother(*H.As<HypreParMatrix>());
+      dynamic_cast<HypreSmoother *>(HInvPC)->SetType(HypreSmoother::Jacobi, 1);
+   }
+   HInv = new CGSolver(MPI_COMM_WORLD);
+   HInv->iterative_mode = true;
+   HInv->SetOperator(*H);
+   HInv->SetPreconditioner(*HInvPC);
+   HInv->SetPrintLevel(pl_hsolve);
+   HInv->SetRelTol(rtol_hsolve);
+   HInv->SetMaxIter(200);
+
+   un_gf.GetTrueDofs(un);
+
+   sw_setup.Stop();
+}
+
+void NavierSolver::Step(double &time, double dt, int cur_step)
+{
+   if (verbose && pmesh->GetMyRank() == 0)
+   {
+      mfem::out << "Step" << std::endl;
+   }
+   sw_step.Start();
+
+   time += dt;
+
+   // Set current time for velocity dirichlet boundary conditions.
+   for (auto &vel_dbc : vel_dbcs)
+   {
+      vel_dbc.coeff.SetTime(time);
+   }
+
+   // Set current time for pressure dirichlet boundary conditons.
+   for (auto &pres_dbc : pres_dbcs)
+   {
+      pres_dbc.coeff.SetTime(time);
+   }
+
+   SetTimeIntegrationCoefficients(cur_step);
+
+   if (cur_step <= 2)
+   {
+      H_bdfcoeff.constant = bd0 / dt;
+      H_form->Update();
+      H_form->Assemble();
+      H_form->FormSystemMatrix(vel_ess_tdof, H);
+
+      if (partial_assembly)
+      {
+         // @TODO: Do I need to clear the old preconditioner? Because
+         // SetOperator does not do that.
+         // HInv->ClearPreconditioner();
```

Comment:
I guess `ClearPreconditioner` didn't make it into this PR. This has the consequence of calling `SetOperator` on the preconditioner, which can sometimes be expensive. In the case of OperatorJacobi it shouldn't matter at all